File: php/test/CloneEntityTest.php
<?php
declare(strict_types=1);

// Clone entity test

require_once __DIR__ . '/../bluefinshieldconexmgmt_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class CloneEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = BluefinShieldconexMgmtSDK::test(null, null);
        $ent = $testsdk->Clone(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = clone_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "clone." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $clone_ref01_ent = $client->Clone(null);
        $clone_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.clone"), "clone_ref01"));
        $clone_ref01_data["template_id"] = $setup["idmap"]["template01"];

        $clone_ref01_data_result = $clone_ref01_ent->create($clone_ref01_data, null);
        $clone_ref01_data = Helpers::to_map($clone_ref01_data_result);
        $this->assertNotNull($clone_ref01_data);
        $this->assertNotNull($clone_ref01_data["id"]);

    }
}

function clone_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/clone/CloneTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = BluefinShieldconexMgmtSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["clone01", "clone02", "clone03", "template01", "template02", "template03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID" => $idmap,
        "BLUEFINSHIELDCONEXMGMT_TEST_LIVE" => "FALSE",
        "BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN" => "FALSE",
        "BLUEFINSHIELDCONEXMGMT_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["BLUEFINSHIELDCONEXMGMT_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new BluefinShieldconexMgmtSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: cpp/test/clone_entity_test.cpp
// Generated basic-flow test for the clone entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct CloneSetup {
  std::shared_ptr<BluefinShieldconexMgmtSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static CloneSetup clone_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/clone/CloneTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinShieldconexMgmtSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("clone01"), Value("clone02"), Value("clone03"), Value("template01"), Value("template02"), Value("template03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID", idmap},
    {"BLUEFINSHIELDCONEXMGMT_TEST_LIVE", Value("FALSE")},
    {"BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFINSHIELDCONEXMGMT_TEST_LIVE") == Value("TRUE");

  CloneSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void clone_entity_instance() {
  auto testsdk = BluefinShieldconexMgmtSDK::testSDK();
  auto ent = testsdk->clone();
  ASSERT_EQ(ent->getName(), std::string("clone"), "entity name");
}

static void clone_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"clone", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = BluefinShieldconexMgmtSDK::testSDK(seed, sdkopts);
  auto se = strsdk->clone();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = BluefinShieldconexMgmtSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->clone();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void clone_entity_basic() {
  auto setup = clone_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create"}) {
    auto sk = is_control_skipped("entityOp", std::string("clone.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto clone_ref01_ent = client->clone();
  Value clone_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "clone"}), "clone_ref01"));
  if (!clone_ref01_data.is_map()) clone_ref01_data = vmap();
  setp(clone_ref01_data, "template_id", getp(setup.idmap, "template01"));
  {
    Value clone_ref01_data_result = clone_ref01_ent->create(Struct::clone(clone_ref01_data), Value::undef());
    clone_ref01_data = Helpers::toMapAny(clone_ref01_data_result);
    if (!clone_ref01_data.is_map()) clone_ref01_data = vmap();
    ASSERT_TRUE(clone_ref01_data.is_map(), "expected create result to be a map");
    ASSERT_TRUE(!getp(clone_ref01_data, "id").is_undef(), "expected created entity to have an id");
  }

}

int main() {
  T_RUN(clone_entity_instance);
  T_RUN(clone_entity_stream);
  T_RUN(clone_entity_basic);
  return sdktest::summary("clone_entity_test");
}
